### PR TITLE
Fix the flush_status bug in flush-executor

### PR DESCRIPTION
**Pull request by lingbin**
For a tablet, there may be multiple memtables, which will be
flushed to disk one by one in the order of generation.

If a memtable flush fails, then the load job will definitely
fail, but the previous implementation will overwrite `_flush_status`,
which may make the error can not be detected, leads to an error
load job to be success.

This patch also have two other changes:
1. Use `std::bind` to replace `boost::bind`;
2. Removes some unneeded headers.

this is an TODO in #2931

### CODE REVIEW

File: be/src/olap/memtable_flush_executor.cpp [hunk 1]
```diff
@@ -35,15 +31,15 @@ std::ostream& operator<<(std::ostream& os, const FlushStatistic& stat) {
 }
 
 OLAPStatus FlushToken::submit(std::shared_ptr<MemTable> memtable) {
```
Comment (imay) [inline be/src/olap/memtable_flush_executor.cpp]: why not use `const std::shared_ptr<MemTable>&` to avoid a copy assignment?

Comment (lingbin) [reply be/src/olap/memtable_flush_executor.cpp]: After the `submit()` returns, the upper layer will immediately reset the `memtable` (let it be `nullptr`), and the lower thread pool may not yet start processing it. If a reference is used, when the thread pool starts to process this `memtable`, the corresponding object may have become invalid.

Comment (imay) [reply be/src/olap/memtable_flush_executor.cpp]: `std::bind` below copies `memtable`, which will keep it valid in your mentioned situation. So I think it is OK to make the function's argument reference.

Comment (lingbin) [reply be/src/olap/memtable_flush_executor.cpp]: Oh, I get you. I will change the param to a reference.

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**